### PR TITLE
Simplify irrefutable slice patterns

### LINKED ISSUE #47096: ICE: match pairs remaining after simplifying irrefutable pattern
Irrefutable slice patterns with a `binding..` may cause the compiler to ICE.

```rust
#![feature(slice_patterns)]

fn foo(&[ref _xs..]: &[i32]) {}

fn main() {}
```

I expected to see this happen: the code compiled (however silly this example is)

Instead, this happened:

```
error: internal compiler error: /checkout/src/librustc_mir/build/matches/mod.rs:177: match pairs [MatchPair { place: (*_1), pattern: Pattern { ty: [i32], span: slice_patterns_ice.rs:3:9: 3:20, kind: Slice { prefix: [], slice: Some(Pattern { ty: [i32], span: slice_patterns_ice.rs:3:10: 3:17, kind: Binding { mutability: Not, name: _xs, mode: ByRef(ReEmpty, Shared), var: NodeId(9), ty: &'<empty> [i32], subpattern: None } }), suffix: [] } }, slice_len_checked: false }] remaining after simplifying irrefutable pattern
 --> slice_patterns_ice.rs:3:9
  |
3 | fn foo(&[ref _xs..]: &[i32]) {}
  |         ^^^^^^^^^^^

note: the compiler unexpectedly panicked. this is a bug.

note: we would appreciate a bug report: https://github.com/rust-lang/rust/blob/master/CONTRIBUTING.md#bug-reports

note: rustc 1.24.0-nightly (3f916bd30 2017-12-30) running on x86_64-unknown-linux-gnu

thread 'rustc' panicked at 'Box<Any>', /checkout/src/librustc_errors/lib.rs:451:9
note: Run with `RUST_BACKTRACE=1` for a backtrace.
```

## Meta

`rustc --version --verbose`:

rustc 1.24.0-nightly (3f916bd30 2017-12-30)
binary: rustc
commit-hash: 3f916bd3029262e2270dfbafb9ab045927499abd
commit-date: 2017-12-30
host: x86_64-unknown-linux-gnu
release: 1.24.0-nightly
LLVM version: 4.0

Backtrace:

```
   0: std::sys::unix::backtrace::tracing::imp::unwind_backtrace
             at /checkout/src/libstd/sys/unix/backtrace/tracing/gcc_s.rs:49
   1: std::sys_common::backtrace::print
             at /checkout/src/libstd/sys_common/backtrace.rs:68
             at /checkout/src/libstd/sys_common/backtrace.rs:57
   2: std::panicking::default_hook::{{closure}}
             at /checkout/src/libstd/panicking.rs:381
   3: std::panicking::default_hook
             at /checkout/src/libstd/panicking.rs:391
   4: std::panicking::rust_panic_with_hook
             at /checkout/src/libstd/panicking.rs:577
   5: std::panicking::begin_panic
   6: rustc_errors::Handler::span_bug
   7: rustc::session::opt_span_bug_fmt::{{closure}}
   8: rustc::ty::context::tls::with_opt
   9: rustc::session::span_bug_fmt
  10: rustc_mir::build::matches::<impl rustc_mir::build::Builder<'a, 'gcx, 'tcx>>::place_into_pattern
  11: rustc_mir::build::construct_fn
  12: rustc::ty::context::tls::enter
  13: rustc::infer::InferCtxtBuilder::enter
  14: rustc_mir::build::mir_build
  15: rustc_mir::transform::mir_built
  16: rustc::ty::maps::<impl rustc::ty::maps::queries::mir_built<'tcx>>::compute_result
  17: rustc::dep_graph::graph::DepGraph::with_task_impl
  18: rustc_errors::Handler::track_diagnostics
  19: rustc::ty::maps::plumbing::<impl rustc::ty::context::TyCtxt<'a, 'gcx, 'tcx>>::cycle_check
  20: rustc::ty::maps::<impl rustc::ty::maps::queries::mir_built<'tcx>>::force
  21: rustc::ty::maps::<impl rustc::ty::maps::queries::mir_built<'tcx>>::try_get
  22: rustc::ty::maps::TyCtxtAt::mir_built
  23: rustc::ty::maps::<impl rustc::ty::context::TyCtxt<'a, 'tcx, 'lcx>>::mir_built
  24: rustc_mir::transform::check_unsafety::unsafety_check_result
  25: rustc::ty::maps::<impl rustc::ty::maps::queries::unsafety_check_result<'tcx>>::compute_result
  26: rustc::dep_graph::graph::DepGraph::with_task_impl
  27: rustc_errors::Handler::track_diagnostics
  28: rustc::ty::maps::plumbing::<impl rustc::ty::context::TyCtxt<'a, 'gcx, 'tcx>>::cycle_check
  29: rustc::ty::maps::<impl rustc::ty::maps::queries::unsafety_check_result<'tcx>>::force
  30: rustc::ty::maps::<impl rustc::ty::maps::queries::unsafety_check_result<'tcx>>::try_get
  31: rustc::ty::maps::TyCtxtAt::unsafety_check_result
  32: rustc::ty::maps::<impl rustc::ty::context::TyCtxt<'a, 'tcx, 'lcx>>::unsafety_check_result
  33: rustc_mir::transform::mir_const
  34: rustc::ty::maps::<impl rustc::ty::maps::queries::mir_const<'tcx>>::compute_result
  35: rustc::dep_graph::graph::DepGraph::with_task_impl
  36: rustc_errors::Handler::track_diagnostics
  37: rustc::ty::maps::plumbing::<impl rustc::ty::context::TyCtxt<'a, 'gcx, 'tcx>>::cycle_check
  38: rustc::ty::maps::<impl rustc::ty::maps::queries::mir_const<'tcx>>::force
  39: rustc::ty::maps::<impl rustc::ty::maps::queries::mir_const<'tcx>>::try_get
  40: rustc::ty::maps::TyCtxtAt::mir_const
  41: rustc::ty::maps::<impl rustc::ty::context::TyCtxt<'a, 'tcx, 'lcx>>::mir_const
  42: rustc_mir::transform::mir_validated
  43: rustc::ty::maps::<impl rustc::ty::maps::queries::mir_validated<'tcx>>::compute_result
  44: rustc::dep_graph::graph::DepGraph::with_task_impl
  45: rustc_errors::Handler::track_diagnostics
  46: rustc::ty::maps::plumbing::<impl rustc::ty::context::TyCtxt<'a, 'gcx, 'tcx>>::cycle_check
  47: rustc::ty::maps::<impl rustc::ty::maps::queries::mir_validated<'tcx>>::force
  48: rustc::ty::maps::<impl rustc::ty::maps::queries::mir_validated<'tcx>>::try_get
  49: rustc::ty::maps::TyCtxtAt::mir_validated
  50: rustc::ty::maps::<impl rustc::ty::context::TyCtxt<'a, 'tcx, 'lcx>>::mir_validated
  51: rustc_borrowck::borrowck::borrowck
  52: rustc::dep_graph::graph::DepGraph::with_task_impl
  53: rustc_errors::Handler::track_diagnostics
  54: rustc::ty::maps::plumbing::<impl rustc::ty::context::TyCtxt<'a, 'gcx, 'tcx>>::cycle_check
  55: rustc::ty::maps::<impl rustc::ty::maps::queries::borrowck<'tcx>>::force
  56: rustc::ty::maps::<impl rustc::ty::maps::queries::borrowck<'tcx>>::try_get
  57: rustc::ty::maps::TyCtxtAt::borrowck
  58: rustc::ty::maps::<impl rustc::ty::context::TyCtxt<'a, 'tcx, 'lcx>>::borrowck
  59: rustc_borrowck::borrowck::check_crate
  60: <std::thread::local::LocalKey<T>>::with
  61: <std::thread::local::LocalKey<T>>::with
  62: rustc::ty::context::TyCtxt::create_and_enter
  63: rustc_driver::driver::compile_input
  64: rustc_driver::run_compiler

```

### PR DESCRIPTION
**Pull request by topecongiro**
Closes #47096.

### CODE REVIEW

**Comment by topecongiro** [comment]
Fixed test failures & rebased.

File: src/librustc_mir/build/matches/simplify.rs [hunk 1]
```diff
@@ -92,11 +92,19 @@ impl<'a, 'gcx, 'tcx> Builder<'a, 'gcx, 'tcx> {
...
-            PatternKind::Range { .. } |
-            PatternKind::Slice { .. } => {
+            PatternKind::Range { .. } => {
                 Err(match_pair)
             }
 
+            PatternKind::Slice { ref prefix, ref slice, ref suffix } => {
+                if prefix.is_empty() && slice.is_some() && suffix.is_empty() {
+                    // irrefutable
+                    Ok(())
```
Comment (nikomatsakis) [inline src/librustc_mir/build/matches/simplify.rs]: Something seems missing here. Doesn't this produce a binding for the slice? If so, don't we have to extend `candidate.bindings`?

Maybe try an example that actually uses the binding? e.g.

```rust
let [foo..] = s;
println!("{:?}", foo);
```

Comment (topecongiro) [reply src/librustc_mir/build/matches/simplify.rs]: Thank you for you review! I tried the example, and it died with segmentation fault. So yes, we need to extend `candidate.bindings`. I will update this PR to do so.

**Comment by topecongiro** [comment]
Update this PR and rebased.

File: src/test/run-pass/irrefutable-slice-patterns.rs [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+// option. This file may not be copied, modified, or distributed
+// except according to those terms.
+
+// #47096
+
+#![feature(slice_patterns)]
+#![allow(dead_code)]
+
+fn foo(s: &[i32]) {
+    let &[ref _xs..] = s;
```
Comment (nikomatsakis) [inline src/test/run-pass/irrefutable-slice-patterns.rs]: Can we update this test to actually use `xs`?

For example:

```rust
fn foo(s: &[i32]) -> &[i32] {
    let &[ref s1..] = s;
    s1
}

fn main() {
    let x = [1, 2, 3];
    let y = foo(&x);
    assert_eq!(y, &[1, 2, 3]);
}
```

**Review by nikomatsakis — CHANGES_REQUESTED** [review]

File: src/librustc_mir/build/matches/simplify.rs [hunk 3]
```diff
@@ -92,11 +92,22 @@ impl<'a, 'gcx, 'tcx> Builder<'a, 'gcx, 'tcx> {
...
+            PatternKind::Range { .. } => {
                 Err(match_pair)
             }
 
+            PatternKind::Slice { ref prefix, ref slice, ref suffix } => {
+                if prefix.is_empty() && slice.is_some() && suffix.is_empty() {
+                    // irrefutable
+                    candidate
+                        .match_pairs
+                        .push(MatchPair::new(match_pair.place, slice.as_ref().unwrap()));
```
Comment (nikomatsakis) [inline src/librustc_mir/build/matches/simplify.rs]: Huh, so, something here makes me a bit nervous. Why not invoke the helper `prefix_slice_suffix`? For example:


```rust
self.prefix_slice_suffix(
    &mut candidate.match_pairs,
    match_pair.place,
    prefix,
    slice,
    suffix,
);
```

that helper is more general than strictly needed, but I think it is sure to do the right thing, or at least we can do the wrong thing all in the same code. =)

Comment (nikomatsakis) [reply src/librustc_mir/build/matches/simplify.rs]: (In this particular case, the main difference would be that the `Place` would include a `Subslice` projection, I think, which may not be needed, but I still think i'd rather go through the helper.)

Comment (nikomatsakis) [reply src/librustc_mir/build/matches/simplify.rs]: Thoughts?

**Review by nikomatsakis — CHANGES_REQUESTED** [review]

File: src/librustc_mir/build/matches/simplify.rs (hunk 3 above)
Comment (topecongiro) [reply src/librustc_mir/build/matches/simplify.rs]: In `prefix_slice_suffix()`, we have 
```rust
        if let Some(subslice_pat) = opt_slice {
            let subslice = place.clone().elem(ProjectionElem::Subslice {
                from: prefix.len() as u32,
                to: suffix.len() as u32
            });
            match_pairs.push(MatchPair::new(subslice, subslice_pat));
        }
```
so `ProjectionElem::Subslice.to` is set to `suffix.len()` which is `0` in this case.

Doc comment on `Subslice` in librustc/mir/mod.rs says:
```rust
    /// These indices are generated by slice patterns.
    ///
    /// slice[from:-to] in Python terms.
    Subslice {
        from: u32,
        to: u32,
    },
```
In Python, `slice[from:-0]` will be an empty list. This is why I avoided using `prefix_slice_suffix()`. 

However, it turns that setting `to` to `0` is actually equivalent to `slice[from:]` in Python term. So I agree that using `prefix_slice_suffix()` is better here.

File: src/test/run-pass/irrefutable-slice-patterns.rs (hunk 2 above)
Comment (topecongiro) [reply src/test/run-pass/irrefutable-slice-patterns.rs]: Oops, I forgot to do this.

**Review by nikomatsakis — APPROVED** [review]

**Comment by nikomatsakis** [comment]
@bors r+ 

I'm assuming travis will go green here =)

**Comment by kennytm** [comment]
@bors retry 

4 hour timeout due to the [asm.js job](https://travis-ci.org/rust-lang/rust/jobs/332865864) restarting.

**Pull request merged**